src/CreateMiddleware/CreateMiddleware.php

```diff
@@ -0,0 +1,165 @@
+ */
+class CreateMiddleware
+{
+    /**
+     * @var string Template for middleware class.
+     */
+    const CLASS_SKELETON = <<< 'EOS'
+<?php
+
+namespace %namespace%;
+
+use Interop\Http\ServerMiddleware\DelegateInterface;
+use Interop\Http\ServerMiddleware\MiddlewareInterface;
+use Psr\Http\ServerRequestInterface;
+
+class %class% implements MiddlewareInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function process(ServerRequestInterface $request, DelegateInterface $delegate)
+    {
+        // $response = $delegate->process($request);
+    }
+}
+EOS;
+
+    /**
+     * @param string $class
+     * @param string $projectRoot
+     * @return string
+     * @throws CreateMiddlewareException
+     */
+    public function process($class, $projectRoot)
+    {
+        $path = $this->getClassPath($class, $projectRoot);
+
+        list($namespace, $class) = $this->getNamespaceAndClass($class);
+
+        $content = str_replace(
+            ['%namespace%', '%class%'],
+            [$namespace, $class],
+            self::CLASS_SKELETON
+        );
+
+        file_put_contents($path, $content);
+        return $path;
+    }
+
+    /**
+     * @param string $class
+     * @param string $projectRoot
+     * @return string
+     * @throws CreateMiddlewareException
+     */
+    private function getClassPath($class, $projectRoot)
+    {
+        $autoloaders = $this->getComposerAutoloaders($projectRoot);
+        list($namespace, $path) = $this->discoverNamespaceAndPath($class, $autoloaders);
+
+        // Absolute path to namespace
+        $path = implode([$projectRoot, DIRECTORY_SEPARATOR, $path]);
+
+        $parts = explode('\\', $class);
+        $className = array_pop($parts);
+
+        // Create absolute path to subnamespace, if required
+        $nsParts = explode('\\', trim($namespace, '\\'));
+        $subNsParts = array_slice($parts, count($nsParts));
+
+        if (0 < count($subNsParts)) {
+            $subNsPath = implode(DIRECTORY_SEPARATOR, $subNsParts);
+            $path = implode([$path, DIRECTORY_SEPARATOR, $subNsPath]);
+        }
+
+        // Create path if it does not exist
+        if (! is_dir($path)) {
+            if (false === mkdir($path, 0755, true)) {
+                throw CreateMiddlewareException::unableToCreatePath($path, $class);
+            }
+        }
+
+        return $path . DIRECTORY_SEPARATOR . $className . '.php';
+    }
+
+    /**
+     * @param string $projectRoot
+     * @return array Associative array of namespace/path pairs
+     * @throws CreateMiddlewareException
+     */
+    private function getComposerAutoloaders($projectRoot)
+    {
+        $composerPath = sprintf('%s/composer.json', $projectRoot);
+        if (! file_exists($composerPath)) {
+            throw CreateMiddlewareException::missingComposerJson();
+        }
+
+        $composer = json_decode(file_get_contents($composerPath), true);
+
+        if (json_last_error() !== \JSON_ERROR_NONE) {
```

Comment:
It's an optimization:

- it ensures we're grabbing the global constant, and not one defined in this namespace.
- it prevents PHP from failing to resolve in the current namespace prior to looking for the constant in the global namespace.